[HACKY-102] Support Postgres Enums

micahhahn opened the pull request:
This PR adds a new TemplateHaskell function to create a bridge between Haskell enums and Postgres enums. 

```sql
CREATE TYPE display_element_type AS ENUM ('labeled', 'blank')
```

```hs
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# OPTIONS -Wno-orphans #-}

import Postgres.Enum

data DisplayElementType
  = Labeled
  | Blank
  deriving (Eq, Show, Ord)

$(generatePGEnum ''DisplayElementType "display_element_type" 
    [ ('Labeled, "labeled")
    , ('Blank, "blank")
    ]
  )
```

This allows us to go directly from the Haskell sum type to the database enum type:

```hs
Postgres.doQuery
      connection
      [Postgres.sql|!
         UPDATE content_creation.scaffolding_flows
         SET display_element = ${Labeled}
      |]
      WebServer.expectCommandSuccess
```

```hs
(x :: (List DisplayElementType)) <-
    Postgres.doQuery
      connection
      [Postgres.sql|!
        SELECT intro_button_middle_step_id
        FROM content_creation.scaffolding_flows
      |]
      WebQuery.expectQuerySuccess
```

### Caveats: 
- The `TypeFamilies` extension is required to set up the instances of [`PGRep`](https://hackage.haskell.org/package/postgresql-typed-0.6.2.1/docs/Database-PostgreSQL-Typed-Dynamic.html#t:PGRep) and [`PGType`](https://hackage.haskell.org/package/postgresql-typed-0.6.2.1/docs/Database-PostgreSQL-Typed-Types.html#t:PGType)
- The `PGType` instance will unavoidably be an orphan instance.  So `{-# OPTIONS -Wno-orphans #-}` will need to be set in the module where `generatePGEnum` is used.

### Maintenance Risks: 
- If we upgrade GHC versions, there is a risk that the TemplateHaskell library API changes and we would need to update the `Enum` module appropriately.  I think this risk is relatively low as we don't use any exotic `TH` features. I tried to leave the code very well commented so any NRIer with Haskell experience could probably do the update. 
- If the `postgresql-typed` instances get modified in a future release we may need to update the generated code as well.  Also low risk in my estimation. 

## :microscope: What do you want the reviewer to check?

Please check that we are guarding against all the following errors at compile time (with clear and understandable error messages):
- The type passed as the first argument is not a data type. 
- The type passed as the first argument contains a record constructor
- The type passed as the first argument contains a constructor with arguments
- Specifying the same PG enum value twice in the mapping (e.g. using "blank" twice in the example above)
- Specifying the same constructor name twice in the mapping
- The second argument to `generatePGEnum` is not the name of a type in PG
- The second argument to `generatePGEnum` is the name of a PG type, but not an enum type (e.g. "bool")
- The second argument to `generatePGEnum` is the name of a PG enum type with no values. 
- There are values in the PG enum not present in the mapping (e.g. not specifying "blank" in the example above)
- There are values in the mapping not present in the PG enum (e.g. adding "other" as a mapping value)
- One of the constructors in the data type is not present in the mapping
- A name specified in the mapping is not a constructor in the data type
- The `TypeFamilies` extension is required in the module that runs `generatePGEnum`.  We should show a helpful error message if it's not enabled.

Finally we also will throw a runtime exception if the database produces an unexpected enum value.  This can only happen if the database and compiled exe are out of sync, so test this would require altering the database enum definition after the Haskell is compiled.

Notes on testing this: 

I was able to get a good test setup working directly in `nri-postgresql` by setting up the vars in the .envrc: 

```
export PGHOST=localhost
export PGPORT=8088
export PGDATABASE=content_creation
export PGUSER=content_creation
export PGPASSWORD=content_creation
```

And adding a `EnumTest` module defined like this:

```
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE QuasiQuotes #-}
{-# LANGUAGE TypeFamilies #-}
{-# OPTIONS -Wno-orphans #-}

module Postgres.EnumTest where

import qualified Postgres
import qualified Postgres.Connection
import qualified Postgres.Settings
import Postgres.Enum
import qualified Prelude
import qualified Environment
import qualified Platform

data DisplayElementType
  = Labeled
  | Blank
  deriving (Eq, Show, Ord)

$(generatePGEnum ''DisplayElementType "display_element_type" 
    [ ('Labeled, "labeled")
    , ('Blank, "blank")
    ]
  )

resultTask :: Result a b -> Task a b
resultTask result =
  case result of 
    Ok vvv -> Task.succeed vvv 
    Err eee -> Task.fail eee

g = do 
  nriSettings <- Environment.decode Postgres.Settings.decoder
  connection <- Postgres.Connection.connectionIO nriSettings
  log <- Platform.silentHandler

  _ <- 
    Postgres.doQuery
      connection
      [Postgres.sql|!
         UPDATE content_creation.scaffolding_flows
         SET display_element = ${Labeled}
      |]
      resultTask
        |> Task.attempt log

  (x :: Result Postgres.Error (List DisplayElementType)) <-
    Postgres.doQuery
      connection
      [Postgres.sql|!
        SELECT display_element
        FROM content_creation.scaffolding_flows
      |]
      resultTask
        |> Task.attempt log

  Prelude.pure x
  ```

stoeffel commented on nri-log-explorer/nri-log-explorer.cabal:
```diff
@@ -29,6 +29,7 @@ source-repository head
 executable log-explorer
   main-is: Main.hs
   other-modules:
+      Modifiable
```
Comment in commit is actually wrong. We can just use the `Functor` instance of `ListWidget` for search.

jwoudenberg-nri commented on nri-log-explorer/src/Modifiable.hs:
```diff
@@ -0,0 +1,60 @@
+    filter,
+    reset,
+  )
+where
```
😻

jwoudenberg-nri commented on nri-log-explorer/src/Main.hs:
```diff
@@ -72,8 +70,16 @@ data Model = Model
-  | HasFilter Text (List Text)
-  | EditFilter (Maybe (Text, List Text)) (Edit.Editor Text Name)
+  | HasFilter (Edit.Editor Text Name)
+  | EditFilter (Undo (Edit.Editor Text Name))
```
😻

jwoudenberg-nri commented on nri-log-explorer/src/Main.hs:
```diff
@@ -54,10 +54,8 @@ data Model = Model
-    allRootSpans :: Data.Vector.Vector RootSpan,
     -- The actual data displayed.
-    filteredRootSpans :: ListWidget.List Name RootSpan,
+    rootSpans :: Modifiable.ListWidget Name RootSpan,
```
😻 love consolidation of these two fields!

jwoudenberg-nri commented on nri-log-explorer/src/Main.hs:
```diff
@@ -764,12 +758,12 @@ handleEvent pushMsg model event =
     (Brick.AppEvent msg) -> update model msg
 
-data Mode = NormalMode | EditMode (Maybe (Text, List Text)) (Edit.Editor Text Name)
+data Mode = NormalMode | EditMode (Undo (Edit.Editor Text Name))
```
😻 Great that this mode type has become simpler! Would be nice if we can eventually figure out a way to turn it into a plain enum I think.

jwoudenberg-nri reviewed (approved):
This is much nicer, thank you!

micahhahn committed:
Expose types to try using enums

micahhahn committed:
PGRep is from Dynamic

micahhahn committed:
All Enum methods

micahhahn committed:
Need PGStringType

micahhahn committed:
Add some basic checking and error messages

micahhahn committed:
Remove expose expirement

BrianHicks commented on nri-postgresql/src/Postgres/Enum.hs:
```diff
@@ -0,0 +1,97 @@
+
+
+quote :: Text -> Text
+quote t = "'" ++ t ++ "'"
```
does this need to escape any `'` inside `t`?

micahhahn replied:
I thiiiink we can trust these names not to have a `'` in them

BrianHicks replied:
doesn't feel safe. I know we don't use `'` in names a lot at NRI but it seems pretty common to prime-ify something to avoid a name collision in the rest of the ecosystem.

micahhahn replied:
Ohhh yeah I was thinking about prefixed `'` cause of all the TemplateHaskell, but I didn't think about postfix quotes

micahhahn committed:
Refactoring + data type checks

micahhahn committed:
Check against duplicated constructor names

micahhahn committed:
Add constructor name validations

micahhahn committed:
Instances!

micahhahn committed:
Remove unneeded modules

micahhahn committed:
Fix up warnings

micahhahn committed:
Use nri connection

micahhahn committed:
Better string quoting

micahhahn committed:
Give the user a better warning if TypeFamilies is not enabled

micahhahn committed:
Improve error message in the unexpected value case

micahhahn commented:
@caiquefigueiredo , @jali-clarke , @omnibs , @stoeffel  I wasn't sure who should be the reviewer of this, so I assigned you all as people who have worked on our haskell-libraries.

micahhahn committed:
Update the docs a bit

BrianHicks commented:
sorry, I'm probably not the right person to give a review on this. The idea sounds extremely cool but I'd be looking at it like "oh yeah, cool cool, yeah I'm sure that's fine." 😆

stoeffel commented:
Is there a reason for not using the modules provided by postgresql-typed?

* https://hackage.haskell.org/package/postgresql-typed-0.6.2.1/docs/Database-PostgreSQL-Typed-Enum.html
* see example https://github.com/dylex/postgresql-typed/blob/master/test/Main.hs#L42

micahhahn commented:
@stoeffel 

> Is there a reason for not using the modules provided by postgresql-typed?

Yes! First of all, it just won't work because the TemplateHaskell in `Database.PostgresSQL.Type.Enum` connects to the database using the library environment variables (`TPG_DB`, `TPG_USER`, etc).  We, however, change the name of these in our wrapper (to `PGDATABASE`, `PGUSER`, etc).  Because I wrote this PR as a new TemplateHaskell function I was able to pull in our custom settings:

https://github.com/NoRedInk/haskell-libraries/blob/8d9bdf063ba015d8847a5f02a5add72e0b7a71b5/nri-postgresql/src/Postgres/Enum.hs#L137-L142

Even if that module did work, we don't expose it as part of our `Postgres` library and it would be set probably too high a complexity barrier to have to separately include `postgresql-typed` as a dependency just to have Enum support. 

There are also a couple things I really didn't like with the API of `Database.PostgreSQL.Type.Enum`: 

Because `dataPGEnum` generates the Haskell datatype itself, it has extremely poor discoverability.  Re-using our example above but with `dataPGEnum`:

`dataPGEnum "DisplayElementType" "display_element_type" ("DisplayElement_" ++)` 

Because the type is generated you would not be able to grep a constructor name you found in the codebase (e.g. `DisplayElement_label`) and find this definition. 

I also really didn't like the naming convention `dataPGEnum` forces you into.  This PR's approach is much more flexible in allowing you to define the data type separately however you want and then just type-checking and generating the bridge code. 

Finally I think the approach of this PR is less "magical" in the sense that if you stumbled upon this code w/o context is should be relatively obvious how it's going to work.  Having the one to one relationship between Haskell constructor and PG enum value explicitly visible explains how the code will convert from one to the other. 

```
$(generatePGEnum ''DisplayElementType "display_element_type" 
    [ ('Labeled, "labeled")
    , ('Blank, "blank")
    ]
  )
```

stoeffel commented on nri-postgresql/src/Postgres/Enum.hs:
```diff
@@ -0,0 +1,242 @@
+
+makeValueList :: List Text -> Text
+makeValueList list = 
+  "[" ++ Text.join ", " (List.map (Text.toList >> quote >> Text.fromList) list) ++ "]"
```
I think we don't need to go from Text to String and back, we can just change quote to take and return Text.

stoeffel replied:
Isn't this the same as `show`?

stoeffel replied:
like `makeConList = Prelude.show`?

micahhahn replied:
That's..... a great point lol. Thanks!

stoeffel commented on nri-postgresql/src/Postgres/Enum.hs:
```diff
@@ -0,0 +1,242 @@
+
+            _ ->
+              Nothing
+        )
```
Maybe simpler?
```suggestion
  list
    |> List.sort 
    |> group
    |> List.filterMap 
        (\items -> 
          items
            |> List.drop 1
            |> List.head
        )
```

stoeffel commented on nri-postgresql/src/Postgres/Enum.hs:
```diff
@@ -0,0 +1,242 @@
+          , " ORDER BY enumsortorder"
+          ])
+        
+      case List.map (\[value] -> pgDecodeRep value) enumRows of 
```
wanna filter map and case on the list here?

micahhahn replied:
Sure, was following the lead of the library - I guess the filterMap isn't that much uglier.

stoeffel reviewed (changes requested):
Looking great! Only a few nit-picks.
And one thing I'd like to change before merging; can we add a test for this? Similar to how the test for the postgresql-typed way of doing this works?

micahhahn committed:
🎩 stoeffel for recognizing show in the wild

micahhahn committed:
🎩 stoeffel - makeValueList is also show

micahhahn committed:
Fix typo

micahhahn committed:
Simplify `findDuplicates`

micahhahn committed:
No incomplete patterns

micahhahn committed:
Add a positive compilation test

micahhahn committed:
Unfortunately we can't use `isExtEnabled` in Q if we want to test it

The standard error message is probably good enough (and even recommends that the user enable `TypeFamilies`)!

micahhahn committed:
Add tests against our failure cases

micahhahn committed:
Add query tests

micahhahn commented:
@stoeffel I was able to add some tests both against the compilation failure expectations, and that the queries work at runtime.

micahhahn committed:
Only expose generatePGEnum

micahhahn committed:
Beef up the documentation

micahhahn commented:
I've beefed up the Haddock documentation too: 

<img width="1335" alt="image" src="https://user-images.githubusercontent.com/12899207/201132631-1b08faf1-0194-457e-b48a-227143432d25.png">

stoeffel commented on nri-postgresql/test/Enum.hs:
```diff
@@ -0,0 +1,34 @@
+{-# LANGUAGE TemplateHaskell #-}
```
😻

stoeffel commented on nri-postgresql/src/Postgres/Enum.hs:
```diff
@@ -58,13 +58,6 @@ generatePGEnum :: TH.Name -> Text -> [(TH.Name, Text)] -> TH.Q [TH.Dec]
 generatePGEnum hsTypeName databaseTypeName mapping = do 
   let hsTypeString = Prelude.show hsTypeName
 
-  typeFamiliesEnabled <- TH.isExtEnabled TH.TypeFamilies
```
Yeah, doesn't really seem necessary.

stoeffel commented on nri-postgresql/src/Postgres/Enum.hs:
```diff
@@ -1,11 +1,34 @@
 {-# LANGUAGE TemplateHaskellQuotes #-}
 
+-- | Provides a functionality for generating a bridge between a Haskell type and a Postgres enum type using Template Haskell.
```
🙇

stoeffel reviewed (approved):
Looks awesome! Thanks!
